prow/gerrit/client/client.go

```diff
@@ -658,15 +658,26 @@ func (h *gerritInstanceHandler) QueryChangesForProject(log logrus.FieldLogger, p
 }
 
 func (h *gerritInstanceHandler) queryChangesForProjectWithoutMetrics(log logrus.FieldLogger, project string, lastUpdate time.Time, rateLimit int, additionalFilters ...string) ([]gerrit.ChangeInfo, error) {
-	var pending []gerrit.ChangeInfo
-
 	var opt gerrit.QueryChangeOptions
 	opt.Query = append(opt.Query, strings.Join(append(additionalFilters, "project:"+project), "+"))
 	opt.AdditionalFields = []string{"CURRENT_REVISION", "CURRENT_COMMIT", "CURRENT_FILES", "MESSAGES", "LABELS"}
 
 	log = log.WithFields(logrus.Fields{"query": opt.Query, "additional_fields": opt.AdditionalFields})
 	var start int
 
+	pending := []gerrit.ChangeInfo{}
+	// Deduplicate changes repeated due to pagination, preserving order, and
+	// keeping the last seen.
+	seenPos := map[int]int{}
+	add := func(ci gerrit.ChangeInfo) {
+		if p, ok := seenPos[ci.Number]; ok {
+			pending[p] = ci
```

Comment:
This keeps the last `ChangeInfo` we see for a given change number, but it puts it in the old index which might affect ordering. I'm not sure if that is actually a concern though.

Reply:
Yeah, I thought about that. If we want to keep the final position, it needs tweaking, and could be a lot heavier. Would work to take first-seen when taking the results in reverse order...

Reply:
Hmm, now that you mention it, you're right that this is actually slightly different behavior than your original dedup attempt. Your original version made it so that the latest index was used (not the first seen index). I.e., if we have `[1 a, 2 b, 1 c]` then the old version would give us `[2 b, 1 c]` while the version in this PR will give us `[1 c, 2 b]`.

🤔